N8N-4143 Variables and expressions for pin data nodes without execution

alexgrozav opened the pull request:
https://linear.app/n8n/issue/N8N-4143/make-node-parameter-variables-and-expressions-available-when-using

alexgrozav committed:
fix: Showing pin data without executing the node only in output pane.

alexgrozav committed:
fix: Updated no data message when previous node not executed.

alexgrozav committed:
feat: Added expression input and evaluation for pin data nodes without execution.

alexgrozav committed:
chore: Fixed linting issues and removed remnant console.log().

alexgrozav committed:
chore: Undone package-lock changes.

alexgrozav committed:
fix: Removed pin data store changes.

alexgrozav committed:
fix: Created a new object using vuex runExecutionData.

ivov commented on packages/editor-ui/src/components/VariableSelector.vue:
```diff
@@ -491,7 +520,7 @@ export default mixins(
 				}
 
-				let tempOutputData;
+				let tempOutputData: IVariableSelectorOption[] | null;
```
This is `undefined` here.

ivov commented on packages/editor-ui/src/components/VariableSelector.vue:
```diff
@@ -297,6 +300,32 @@ export default mixins(
+			 * @param {string} nodeName The name of the node to get the data of
+			 * @param {INodeExecutionData} outputData The data of the run to get the data of
+			 * @param {string} filterText Filter text for parameters
+			 * @param {boolean} [useShort=false] Use short notation
```
Nit: I would add a small example here. Not immediately obvious what `short notation` stands for.

alexgrozav replied:
Added description.

ivov commented on packages/editor-ui/src/components/VariableSelector.vue:
```diff
@@ -502,7 +531,41 @@ export default mixins(
+						}
+					});
+
+					if (pinDataOptions.length > 0) {
```
Out of scope: I find this hard to follow without spending longer, which makes it difficult to properly check the logic. In the future, we should make an effort to refactor this section and ideally this whole 250-line method. Since we are looking to release, now is probably not the time.

alexgrozav replied:
I found it very hard to follow as well. The code there is very complicated and requires deep business-logic understanding. 

I've been doing a lot of trial-and-error work to get to understand what this does and how it interacts with the data.

ivov reviewed (commented):
Thanks for taking this on!

1. If we pin a node, connect the next, reference from next to previous using expression, the expression works. If then we add a third node standalone (unconnected), run this third node, and go back to inspect the expression, it now says `[invalid (workflowRunData[parentNode[0]] is undefined)]`

![image](https://user-images.githubusercontent.com/44588767/179946666-d0a91bce-f157-41a7-b63a-8a5bcffdb373.png)

2. If we pin a node, connect the next, run the next, when opening the NDV or the expression edit modal, there are Vue store state mutation warnings and watcher callback error warnings in the console. These fire multiple times when selecting any option (pindata-based or not) from the variable selector.

<img width="1435" alt="image" src="https://user-images.githubusercontent.com/44588767/179948775-ad231c27-a0d9-4ec1-bb48-5238a16da768.png">

alexgrozav committed:
fix: Fixed bug appearing when adding a new node after executing.

alexgrozav commented:
> Thanks for taking this on!
> 
> 1. If we pin a node, connect the next, reference from next to previous using expression, the expression works. If then we add a third node standalone (unconnected), run this third node, and go back to inspect the expression, it now says `[invalid (workflowRunData[parentNode[0]] is undefined)]`
> 
> ![image](https://user-images.githubusercontent.com/44588767/179946666-d0a91bce-f157-41a7-b63a-8a5bcffdb373.png)

Good catch! Found it as well after more thorough testing. Added a fix for it.

> 2. If we pin a node, connect the next, run the next, when opening the NDV or the expression edit modal, there are Vue store state mutation warnings and watcher callback error warnings in the console. These fire multiple times when selecting any option (pindata-based or not) from the variable selector.
> 
> <img alt="image" width="1435" src="https://user-images.githubusercontent.com/44588767/179948775-ad231c27-a0d9-4ec1-bb48-5238a16da768.png">

Updated. Rebuilding workflow run data using spreading to be able to manipulate it.

ivov committed:
fix: Fix editor-ui build

ivov reviewed (approved):
Confirmed, those cases are fixed. FYI pushed a small build fix.

alexgrozav committed:
feat: Added green node connectors when having pin data output.

alexgrozav committed:
Merge branch 'n8n-4143-make-node-parameter-variables-and' of github.com:n8n-io/n8n into n8n-4143-make-node-parameter-variables-and

alexgrozav committed:
chore: Fixed linting errors.

ivov reviewed (commented):
This does not seem to work:
1. Pin test data to Start, connect to next → connector still grey, no `items` label
2. Pin test data to standalone node, connect to next → connector still grey, no `items` label
3. Pin test data to standalone node, connect to next, keep building chain with every node pinned → all connectors still grey, no `items` label in any node

Reviewed: 7feed12dc

alexgrozav commented on packages/editor-ui/src/store.ts:
```diff
@@ -214,6 +215,8 @@ export const store = new Vuex.Store({
 
 			state.stateIsDirty = true;
+
+			dataPinningEventBus.$emit('pin-data', { [payload.node.name]: payload.data });
```
We also need the number of items. See `total: pinData[nodeName].length`.

alexgrozav committed:
fix: Added pin data eventBus unsubscribe.

alexgrozav commented on packages/editor-ui/src/views/NodeView.vue:
```diff
@@ -2956,6 +2959,33 @@ export default mixins(
+
+					connections.forEach((connection) => {
+						CanvasHelpers.resetConnection(connection);
+					});
```
Updated.

alexgrozav commented on packages/editor-ui/src/views/NodeView.vue:
```diff
@@ -3004,10 +3034,19 @@ export default mixins(
+
+			dataPinningEventBus.$on('unpin-data', (pinData: PinData) => {
+				this.removePinDataConnections(pinData);
+			});
```
1. Done.
2. Because I forgot. 🥲

alexgrozav committed:
fix: Added pin data color check after adding a connection.

ivov reviewed (approved):
Amazing work!
